File: python/conduits/sync_messages.py
from pwrpy.pwrsdk import PWRPY
from pwrpy.pwrwallet import Wallet
from pwrpy.models.Transaction import VidaDataTransaction
from pwrpy.TransactionBuilder import TransactionBuilder
from transaction import Transactions
from dotenv import load_dotenv
import json
import time
import os
load_dotenv()

# Setting up your wallet in the SDK
seed_phrase = os.getenv("SEED_PHRASE")
wallet = Wallet.new(seed_phrase)
# Setting up the rpc api
rpc = PWRPY("https://pwrrpc.pwrlabs.io/")

def handler_messages(txn: VidaDataTransaction):
    try:
        sender = txn.sender
        data_hex = txn.data
        # Assuming data_hex is a hex string starting with '0x'
        data_bytes = bytes.fromhex(data_hex)
        # Converting the hex data to a buffer and then to a UTF-8 string
        data_str = data_bytes.decode('utf-8')
        # Parsing the JSON string into a Python dictionary
        obj = json.loads(data_str)
        # Checking if the transaction contains a "message" field with the specified value
        if 'message' in obj and obj['message'].lower() == "please send me pwr":
            fee_per_byte = rpc.get_fee_per_byte()
            # Building a transfer transaction to send PWR tokens
            transfer_txn = TransactionBuilder.get_transfer_pwr_transaction(
                sender, 1000000000, wallet.get_nonce(), rpc.get_chainId(), wallet.get_address(), fee_per_byte
            )
            # Adding the transfer transaction to the Transactions list for later execution
            Transactions.add(transfer_txn)
            # Printing a message to the console showing the sender and the message content
            print(f"\nMessage from 0x{sender}: {obj['message']}")
    except Exception as e:
        print('Error in sync:', e)
        time.sleep(1)

def sync():
    starting_block = rpc.get_latest_block_number()
    vida_id = 123
    rpc.subscribe_to_vida_transactions(vida_id, starting_block, handler=handler_messages)

File: go/conduits/syncMessages.go
package conduits

import (
    "encoding/hex"
    "encoding/json"
    "fmt"
    "os"
    "github.com/joho/godotenv"
    "github.com/pwrlabs/pwrgo/rpc"
    "github.com/pwrlabs/pwrgo/wallet"
    "github.com/pwrlabs/pwrgo/encode"
)

func handlerMessages(transaction rpc.VidaDataTransaction) {
    godotenv.Load()
    seedPhrase := os.Getenv("SEED_PHRASE")
    wallet, err := wallet.New(seedPhrase)
    if err != nil {
        fmt.Println("Error getting the wallet:", err)
        return
    }

    sender := transaction.Sender
    dataHex := transaction.Data
    nonce := wallet.GetNonce()

    // Converting the hex data to a buffer and then to a UTF-8 string
    dataBytes, _ := hex.DecodeString(dataHex)
    var obj map[string]interface{}
    if err := json.Unmarshal(dataBytes, &obj); err != nil {
        fmt.Println("Error parsing JSON:", err)
    }

    // Iterating over each key in the object to check for specific conditions
    if message, ok := obj["message"].(string); ok && message == "please send me pwr" {
        var buffer []byte
        // Building a transfer transaction to send PWR tokens
        buffer, err := encode.TransferTxBytes(1000000000, sender, nonce, wallet.Address, wallet.GetRpc().GetFeePerByte())
        if err != nil {
            fmt.Println("Error encoding transaction:", err)
        }
        // Adding the transaction to the Transactions struct
        PendingTransactions.Add(buffer)
        // Logging the message and the sender to the console
        fmt.Printf("\nMessage from 0x%s: %s\n", sender, message)
    }
}

func Sync() {
    rpc := rpc.SetRpcNodeUrl("https://pwrrpc.pwrlabs.io")
    startingBlock := rpc.GetLatestBlockNumber()
    vidaId := 123

    rpc.SubscribeToVidaTransactions(vidaId, startingBlock, handlerMessages)
}